promotion.md

```diff
@@ -105,12 +68,45 @@ extension but actually does not:
 const EMPTY_BYTES: &Vec<u8> = &Vec::new(); // Ok without lifetime extension
 ```
 
-As we have seen above, `Vec::new()` does not get promoted. And yet this
+`Vec::new()` cannot get promoted because it needs dropping. And yet this
 compiles. Why that? The reason is that the reference obtains the lifetime of
 the "enclosing scope", similar to how `let x = &mut x;` creates a reference
 whose lifetime lasts for the enclosing scope. This is decided during MIR
 building already, and does not involve lifetime extension.
 
+## Implicit and explicit promotion
+
+On top of what applies to [consts](const.md), promoteds suffer from the additional issue that *the user did not ask for them to be evaluated at compile-time*.
+Thus, if CTFE fails but the code would have worked fine at run-time, we broke the user's code for no good reason.
+Even if we are sure we found an error in the user's code, we are only allowed to [emit a warning, not a hard error][warn-rfc].
+We call this *implicit* promotion, and we have to be very conservative with what can and cannot be implicitly promoted.
+
+CTFE of implicitly promoted code must never fail to evaluate except if the
+run-time code also would have failed. This means we cannot permit calling
+arbitrary `const fn`, as discussed in detail in
+[rust-lang/const-eval#19](https://github.com/rust-lang/const-eval/issues/19).
+Thus, only functions marked `#[rustc_promotable]` are implicitly promotable (see
+below).
+
+On the other hand, when a user passes an expression to a function with
+`#[rustc_args_required_const]`, the only way for this code to compile is to
+promote it.  In that sense, the user is explicitly asking for that expression to
+be evaluated at compile-time even though they have not written it in a `const`
+declaration. We can thus be less conservative. This is called *explicit*
+promotion.
+
+Currently, the following are considered explicit promotion contexts:
+* `#[rustc_args_required_const]` arguments
+* lifetime extension in the bodies of `const` and `static` items and array lengths
+
+Everything else is an implicit promotion context, including `const fn` bodies and non-`Copy` array initializers.
```

Comment:
`const fn` bodies are also explicit promotion contexts:

```rust
const fn foo() -> i32 {
    42
}

const fn bar() -> &'static i32 {
    &foo()
}
```

Reply:
But... that's bad? They can be evaluated at runtime!

Reply:
sure, but you can't do anything bad as the expression must already be const evaluable

Reply:
You can do tons of bad things, like divide by 0 or panic. This breaks everything, it means we cannot make const-eval failures hard errors.  Just imagine doing codegen for
```rust
fn evil() { /* do something nasty */ }

fn bar() {
  if false { &evil() }
}
```
We have to evaluate the promoted for codegen, but if evaluation fails we cannot cause a hard error as that would mean we broke the build for no good reason.

Since control flow has just gotten stable in const, can we quickly fix this hole and make `const fn` *implicit* promotion contexts?

Reply:
> Since control flow has just gotten stable in const, can we quickly fix this hole and make `const fn` _implicit_ promotion contexts?

turning it into implicit promotion contexts is a breaking change that is entirely independent of control flow

Reply:
Sure, but without control flow there's no dead code.

... not sure how that helps mitigate the scope of the breaking change though. But I feel like this is still something we should try ASAP, as otherwise all hope is lost of ever treating promoteds in a sane way. :(

Reply:
This is what happens when we do not write down proper design documents. :( I "naturally" assumed that `const fn` are treated like `fn` for promotion (because they can be evaluated at runtime, and thus all of the problems of promotion in `fn` also apply in `const fn`), and you "naturally" assumed that `const fn` are treated like `const`/`static` initializers. And since I cannot actually understand the code that implements this and we never wrote any of this down at a high level, here we are...

Reply:
Oh no. I wish I would have realized this while doing the initial refactoring of const-checking. That logic is [here](https://github.com/rust-lang/rust/blame/master/src/librustc_mir/transform/promote_consts.rs#L761-L771) BTW.

Reply:
I wish I had read the code more careful to be able to provide such input more timely. No point lamenting over spilled beans, though I wonder if there is something we can do with our process to help with such things, without drowning in paperwork. The MCP mechanism could have helped, assuming that such details of the desired analysis would be spelled out in the MCP.

>  That logic is here BTW

This looks like `const fn` are not an explicit promotion context, but get some kind of special treatment that is half-way between explicit and implicit contexts? Are there other such "hybrid" cases?

> I can do a crater run to see how many people depend on this.

Yes that would be great.

Reply:
Note to self: this has worked for every stable release after 1.35.

Reply:
I have collected my current thoughts on this [in a hackmd](https://hackmd.io/7DekSpq3R-KgizegdAWO3Q). Please feel free to expand that with missing information!

Reply:
I opened an issue at https://github.com/rust-lang/rust/issues/75586, so we have a dedicated place to track this problem. I will adjust the text here to reality.